File: src/main/java/com/extendedae_plus/common/init/ModItems.java
package com.extendedae_plus.common.init;

import appeng.items.materials.UpgradeCardItem;
import appeng.items.parts.PartItem;
import com.extendedae_plus.ExtendedAEPlus;
import com.extendedae_plus.common.block.EAEPCraftingUnitType;
import com.extendedae_plus.common.item.*;
import com.extendedae_plus.common.item.infinityBigIntegerCell.InfinityBigIntegerCellItem;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.Item;
import net.neoforged.neoforge.registries.DeferredBlock;
import net.neoforged.neoforge.registries.DeferredItem;
import net.neoforged.neoforge.registries.DeferredRegister;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;
import java.util.random.RandomGenerator;

public final class ModItems {
    public static final DeferredRegister.Items ITEM =
            DeferredRegister.createItems(ExtendedAEPlus.MODID);

    /// CraftingUnit Items are now in {@link EAEPCraftingUnitType#UNIT_ITEMS}
    public static final List<DeferredItem<?>> ITEMS = new ArrayList<>();

    public static final DeferredItem<BlockItem> WIRELESS_TRANSCEIVER =
            regCommonBlockItem("wireless_transceiver", ModBlocks.WIRELESS_TRANSCEIVER);
    public static final DeferredItem<BlockItem> ASSEMBLER_MATRIX_UPLOAD_CORE =
            regCommonBlockItem("assembler_matrix_upload_core", ModBlocks.ASSEMBLER_MATRIX_UPLOAD_CORE);

    public static final DeferredItem<PartItem<?>> ENTITY_TICKER_PART_ITEM =
            regItem("entity_speed_ticker", EntitySpeedTickerPartItem::new);

    public static final DeferredItem<Item> INFINITY_BIGINTEGER_CELL_ITEM =
            regItem("infinity_biginteger_cell", InfinityBigIntegerCellItem::new);
    public static final DeferredItem<Item> PROVIDER_CONTROLLER =
            regItem("provider_controller", ItemProviderController::new);

    public static final DeferredItem<UpgradeCardItem> CHANNEL_CARD =
            regItem("channel_card", ChannelCardItem::new);
    public static final DeferredItem<UpgradeCardItem> CARD_AUTO_COMPLETION =
            regItem("card_auto_completion", ItemCardAutoCompletion::new);

    /// 随机数, 嘻嘻😋
    public static final DeferredItem<UpgradeCardItem> ENTITY_SPEED_CARD =
            ITEM.register("entity_speed_card", () -> new EntitySpeedCardItem(RandomGenerator.getDefault().nextInt(64)));

    public static <T extends Item> DeferredItem<T> regItem(String name, Supplier<T> factory) {
        var holder = ITEM.register(name, factory);
        ITEMS.add(holder);
        return holder;
    }
    
    public static DeferredItem<BlockItem> regCommonBlockItem(String name, DeferredBlock<?> block) {
        var holder = ITEM.register(name, () -> new BlockItem(block.get(), new Item.Properties()));
        ITEMS.add(holder);
        return holder;
    }
}

File: src/main/kotlin/com/extendedae_plus/common/init/ModItems.kt
package com.extendedae_plus.common.init

import appeng.items.materials.UpgradeCardItem
import com.extendedae_plus.ExtendedAEPlus
import com.extendedae_plus.common.registry.item.ItemProviderController
import com.extendedae_plus.common.registry.item.ItemTicker
import com.extendedae_plus.common.registry.item.infinityBigIntegerCell.InfinityBigIntegerCellItem
import com.extendedae_plus.common.registry.item.priorityTool.ItemPriorityTool
import com.extendedae_plus.common.registry.item.upgradeCard.CardAutoCompletion
import com.extendedae_plus.common.registry.item.upgradeCard.CardChannel
import com.extendedae_plus.common.registry.item.upgradeCard.CardTicking
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.neoforged.neoforge.registries.DeferredBlock
import net.neoforged.neoforge.registries.DeferredItem
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Supplier
import java.util.random.RandomGenerator

object ModItems {
    @InitObject
    val Register: DeferredRegister.Items = DeferredRegister.createItems(ExtendedAEPlus.MODID)

    /** CraftingUnit Items are now in [com.extendedae_plus.common.registry.block.EAEPCraftingUnitType.UNIT_ITEMS] */
    val Items = ArrayList<DeferredItem<*>>()

    @JvmField
    val WirelessTransceiver = this.regCommonBlockItem(ModBlocks.WirelessTransceiver)
    @JvmField
    val PortUpload = this.regCommonBlockItem(ModBlocks.PortUpload)
    @JvmField
    val CoreAdvancedCrafter = this.regCommonBlockItem(ModBlocks.CoreAdvancedCrafter)
    @JvmField
    val CoreAdvancedPattern = this.regCommonBlockItem(ModBlocks.CoreAdvancedPattern)
    @JvmField
    val CoreAdvancedSpeed = this.regCommonBlockItem(ModBlocks.CoreAdvancedSpeed)

    @JvmField
    val Ticker = this.regItem("ticker", ::ItemTicker)

    @JvmField
    val CellInfinity = this.regItem("infinity_biginteger_cell", ::InfinityBigIntegerCellItem)
    @JvmField
    val ControllerProvider = this.regItem("provider_controller", ::ItemProviderController)
    @JvmField
    val PriorityTool = this.regItem("priority_tool", ::ItemPriorityTool)

    @JvmField
    val CardChannel = this.regItem("channel_card", ::CardChannel)
    @JvmField
    val CardAutoCompletion = this.regItem("card_auto_completion", ::CardAutoCompletion)

    /** 随机数, 嘻嘻😋 */
    @JvmField
    val CardTicking: DeferredItem<UpgradeCardItem> = Register.register(
        "card_ticking",
        Supplier { CardTicking(RandomGenerator.getDefault().nextInt(64), 114514) }
    )

    fun <T : Item> regItem(name: String, constructor: () -> T): DeferredItem<T> {
        val holder = Register.register(name, constructor)
        Items.add(holder)
        return holder
    }

    fun regCommonBlockItem(block: DeferredBlock<*>): DeferredItem<BlockItem> =
        this.regCommonBlockItem(block.id.path, block)

    fun regCommonBlockItem(name: String, block: DeferredBlock<*>): DeferredItem<BlockItem> =
        this.regItem(name) { BlockItem(block.get(), Item.Properties()) }
}
